src/main/java/net/iponweb/disthene/service/aggregate/AggregateService.java

```diff
@@ -0,0 +1,179 @@
+package net.iponweb.disthene.service.aggregate;
+
+import com.google.common.util.concurrent.AtomicDouble;
+import com.google.common.util.concurrent.RateLimiter;
+import net.engio.mbassy.bus.MBassador;
+import net.engio.mbassy.listener.Handler;
+import net.engio.mbassy.listener.Listener;
+import net.engio.mbassy.listener.References;
+import net.iponweb.disthene.bean.Metric;
+import net.iponweb.disthene.bean.MetricKey;
+import net.iponweb.disthene.config.DistheneConfiguration;
+import net.iponweb.disthene.config.Rollup;
+import net.iponweb.disthene.events.DistheneEvent;
+import net.iponweb.disthene.events.MetricAggregateEvent;
+import net.iponweb.disthene.events.MetricStoreEvent;
+import net.iponweb.disthene.util.NamedThreadFactory;
+import org.apache.log4j.Logger;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * @author 
+ */
+@Listener(references= References.Strong)
+public class AggregateService {
+    private static final String SCHEDULER_NAME = "distheneAggregatorFlusher";
+    private static final int RATE = 10;
+    private volatile boolean shuttingDown = false;
+
+    private Logger logger = Logger.getLogger(AggregateService.class);
+
+    private MBassador<DistheneEvent> bus;
+    private DistheneConfiguration distheneConfiguration;
+    private Rollup baseRollup;
+
+    private ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1, new NamedThreadFactory(SCHEDULER_NAME));
+
+    private final ConcurrentNavigableMap<Long, ConcurrentMap<MetricKey, AverageRecord>> accumulator = new ConcurrentSkipListMap<>();
+
+    public AggregateService(MBassador<DistheneEvent> bus, DistheneConfiguration distheneConfiguration) {
+        this.distheneConfiguration = distheneConfiguration;
+        this.baseRollup = distheneConfiguration.getCarbon().getBaseRollup();
+        this.bus = bus;
+        bus.subscribe(this);
+
+        scheduler.scheduleAtFixedRate(new Runnable() {
+            @Override
+            public void run() {
+                flush();
+            }
+        }, 60 - ((System.currentTimeMillis() / 1000L) % 60), RATE, TimeUnit.SECONDS);
+    }
+
+    @Handler(rejectSubtypes = false)
+    public void handle(MetricAggregateEvent metricAggregateEvent) {
+        aggregate(metricAggregateEvent.getMetric());
+    }
+
+    private ConcurrentMap<MetricKey, AverageRecord> getTimestampMap(long timestamp) {
```

Comment:
It will be nice to have a docstring here

Reply:
fixed